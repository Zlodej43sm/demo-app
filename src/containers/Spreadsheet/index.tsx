import React, { useState, useEffect, ComponentType } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useIntl } from "react-intl";
import {
  Button,
  TableContainer,
  Table,
  TableHead,
  TableRow,
  TableBody,
  TableCell,
} from "@mui/material";

import {
  HEADER_COLUMNS_LIST,
  SERVICE_TEXT,
  PROGRESS_STATUS,
} from "../../common/constants";
import {
  selectSpreadsheetData,
  selectSpreadsheetError,
  selectSpreadsheetProgressStatus,
} from "../../selectors";
import { AppDispatch } from "../../redux/store";
import { fetchData, getStatus } from "../../redux/SpreadsheetSlice";
import SearchInput from "../../common/components/SearchInput";
import CellInput, {
  CurrentSelectedRowProp,
} from "../../common/components/CellInput";
import useStyles from "./styles";

const Spreadsheet: ComponentType = () => {
  const classes = useStyles();
  const dispatch = useDispatch<AppDispatch>();
  const [currentEditRow, setCurrentEditRow] =
    useState<CurrentSelectedRowProp>(null);
  const spreadsheetData = useSelector(selectSpreadsheetData);
  const isError = useSelector(selectSpreadsheetError);
  const currentProgressStatus = useSelector(selectSpreadsheetProgressStatus);
  const { formatMessage } = useIntl();
  const getActiveRowClass = (rowIndex: number): string => {
    let rowClassName = currentEditRow === rowIndex ? classes.selected : "";

    if (isError && currentProgressStatus?.rowIndex === rowIndex) {
      rowClassName = classes.error;
    }

    return rowClassName;
  };

  useEffect(() => {
    dispatch(fetchData());
  }, [dispatch]);

  useEffect(() => {
    const { status, done_at, id } = currentProgressStatus;
    let getStatusTimeout: any;
    if (PROGRESS_STATUS.IN_PROGRESS === status && id && done_at) {
      const doneAtTs = new Date(done_at).getTime();
      const nowTs = new Date().getTime();
      // @TODO: dont ask why, just because fast and demo usage
      const diffTs = Math.abs(doneAtTs - nowTs);
      getStatusTimeout = setTimeout(() => {
        dispatch(getStatus({ id }));
      }, diffTs);
    }
    return () => {
      clearTimeout(getStatusTimeout);
    };
  }, [dispatch, currentProgressStatus]);

  const onAddRowHandler = () => {
    alert(SERVICE_TEXT);
  };
  const onFilterSearchHandler = () => {
    alert(SERVICE_TEXT);
  };

  return (
    <TableContainer
      classes={{ root: classes.tableContainerRoot }}
      style={{ overflowX: "initial" }}
    >
      <Table>
        <TableHead classes={{ root: classes.tableHeadRoot }}>
          <TableRow>
            <TableCell
              classes={{ root: classes.tableSearchBarCellRoot }}
              colSpan={3}
            >
              <SearchInput onSearch={onFilterSearchHandler} />
            </TableCell>
          </TableRow>
          <TableRow classes={{ root: classes.tableHeadRowRoot }}>
            {HEADER_COLUMNS_LIST.map((colValue) => (
              <TableCell
                key={`col-${colValue}`}
                align="center"
                classes={{ root: classes.tableCellRoot }}
              >
                {colValue}
              </TableCell>
            ))}
          </TableRow>
        </TableHead>
        <TableBody classes={{ root: classes.tableBodyRoot }}>
          {spreadsheetData.map((row, rowIndex) => (
            <TableRow
              classes={{ root: classes.tableRowRoot }}
              className={getActiveRowClass(rowIndex)}
              key={`row-${rowIndex}`}
            >
              {row.map(({ id, value, formula }, colIndex) => (
                <TableCell
                  key={`row-${colIndex}`}
                  align="center"
                  classes={{ root: classes.tableCellRoot }}
                >
                  <CellInput id={id} setCurrentEditRow={setCurrentEditRow} />
                </TableCell>
              ))}
            </TableRow>
          ))}
        </TableBody>
      </Table>
      <Button
        variant="outlined"
        onClick={onAddRowHandler}
        classes={{ root: classes.addRowButtonRoot }}
      >
        {formatMessage({ id: "button.add.row" })}
      </Button>
    </TableContainer>
  );
};

export default Spreadsheet;
