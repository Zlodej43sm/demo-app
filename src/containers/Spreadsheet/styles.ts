import { makeStyles } from "@mui/styles";

export const COMPONENT_NAME = "Spreadsheet";

const selectedStyles = (color: string) => ({
  position: "relative",
  zIndex: 5,
  cursor: "pointer",
  borderColor: "transparent",
  outline: "1px solid transparent",
  borderRadius: 5,
  boxShadow: `-1px 0 5px ${color}`,
});

export default makeStyles(
  (theme: any) => ({
    tableContainerRoot: {
      paddingTop: theme.spacing(0.5),
      maxWidth: 615,
      minWidth: 615,

      "& $tableSearchBarCellRoot": {
        padding: 0,
        border: "0 none",
      },
      "& $tableCellRoot": {
        ...theme.typography.body1,
        color: theme.palette.primary.main,
        padding: theme.spacing(),
        border: "0 none",
      },

      "& $addRowButtonRoot": {
        marginRight: "auto",
        marginLeft: "auto",
        marginTop: theme.spacing(1.25),
      },
    },
    tableHeadRoot: {
      top: theme.spacing(1.25),
      position: "sticky",
      zIndex: 3,
    },
    tableHeadRowRoot: {
      borderBottom: `${theme.spacing(1.25)} solid white`,
      "& $tableCellRoot": {
        backgroundColor: theme.palette.grey["50"],
      },
      "& $tableCellRoot:first-child": {
        borderTopLeftRadius: 4,
        borderBottomLeftRadius: 4,
      },
      "& $tableCellRoot:last-child": {
        borderTopRightRadius: 4,
        borderBottomRightRadius: 4,
      },
    },
    tableBodyRoot: {
      "& $tableRowRoot:not($error) $tableCellRoot": {
        backgroundColor: theme.palette.grey["10"],
      },
      "& $tableRowRoot": {
        border: "3px solid white",
        outline: "1px solid transparent",
        "&$selected": selectedStyles(theme.palette.grey["200"]),
        "&:hover:not($error)": selectedStyles(theme.palette.grey["200"]),
      },
      "& $error": selectedStyles(theme.palette.error.main),
    },
    tableCellRoot: {
      "&:not(:last-of-type)": {
        width: "30%",
      },
      "&:last-of-type": {
        width: "40%",
      },
    },
    tableRowRoot: {},
    tableSearchBarCellRoot: {},
    tableCellHead: {},
    addRowButtonRoot: {},
    selected: {},
    error: {
      "& td": {
        backgroundColor: theme.palette.error.light,
      },
    },
  }),
  { name: COMPONENT_NAME }
);
