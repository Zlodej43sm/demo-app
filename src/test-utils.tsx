import React, { ComponentType, PropsWithChildren } from "react";
import { createIntl, RawIntlProvider } from "react-intl";
import { render, RenderOptions } from "@testing-library/react";
import { ThemeProvider, createTheme } from "@mui/material/styles";

import { LANG_CODES } from "./common/constants";
import { appMessages } from "./lang";

const theme = createTheme({
  palette: {},
});
const intl = createIntl({
  locale: LANG_CODES.EN,
  messages: appMessages(LANG_CODES.EN),
});

const AllTheProviders: ComponentType<PropsWithChildren> = ({
  children,
}): any => (
  <ThemeProvider theme={theme}>
    <RawIntlProvider value={intl}>{children}</RawIntlProvider>
  </ThemeProvider>
);

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, "queries">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";

const mockAppStore = {
  spreadsheet: {
    data: [
      [
        {
          id: "A0",
          value: "3",
          formula: "",
        },
        {
          id: "B0",
          value: "18",
          formula: "",
        },
        {
          id: "C0",
          value: 13.5,
          formula: "=12 + 3 / (2)",
        },
      ],
    ],
    byId: {
      A0: [0, 0],
      B0: [0, 1],
      C0: [0, 2],
    },
    progress: {},
    loading: false,
    error: false,
  },
};

export { customRender as render, mockAppStore };
