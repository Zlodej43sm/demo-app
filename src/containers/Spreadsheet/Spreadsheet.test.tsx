import React from "react";
import { useDispatch, useSelector } from "react-redux";

import { render, mockAppStore } from "../../test-utils";
import Spreadsheet from ".";

jest.mock("react-redux", () => ({
  useSelector: jest.fn(),
  useDispatch: jest.fn(),
}));

describe("Spreadsheet", () => {
  const mockedUseSelector = useSelector as jest.Mock<typeof useSelector>;
  const mockedDispatch = useDispatch as jest.Mock<typeof useDispatch>;
  const mockStore = {
    ...mockAppStore,
  };

  beforeEach(() => {
    mockedDispatch.mockReturnValue(jest.fn());
    mockedUseSelector.mockImplementation((selector) => selector(mockStore));
  });

  it("SHOULD match snapshot", () => {
    const { container } = render(<Spreadsheet />);
    expect(container).toMatchSnapshot();
  });
});
