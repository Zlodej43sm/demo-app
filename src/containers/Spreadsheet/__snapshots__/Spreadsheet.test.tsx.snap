// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spreadsheet SHOULD match snapshot 1`] = `
<div>
  <div
    class="MuiTableContainer-root Spreadsheet-tableContainerRoot-1 css-rorn0c-MuiTableContainer-root"
    style="overflow-x: initial;"
  >
    <table
      class="MuiTable-root css-rqglhn-MuiTable-root"
    >
      <thead
        class="MuiTableHead-root Spreadsheet-tableHeadRoot-2 css-15wwp11-MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root Spreadsheet-tableSearchBarCellRoot-7 MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            colspan="3"
            scope="col"
          >
            <div
              class="SearchInput-root-12 MuiBox-root css-0"
            >
              <div
                class="MuiInputBase-root MuiInput-root SearchInput-inputWrapperRoot-14 MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-adornedStart css-1s97tdc-MuiInputBase-root-MuiInput-root"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall SearchInput-searchIconRoot-13 css-ptiqhd-MuiSvgIcon-root"
                  data-testid="SearchIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                  />
                </svg>
                <input
                  class="MuiInputBase-input MuiInput-input MuiInputBase-inputTypeSearch MuiInputBase-inputAdornedStart css-c63i49-MuiInputBase-input-MuiInput-input"
                  placeholder="Type a search query to filter"
                  type="search"
                  value=""
                />
              </div>
            </div>
          </th>
        </tr>
        <tr
          class="MuiTableRow-root Spreadsheet-tableHeadRowRoot-3 MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root Spreadsheet-tableCellRoot-5 MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
            scope="col"
          >
            A
          </th>
          <th
            class="MuiTableCell-root Spreadsheet-tableCellRoot-5 MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
            scope="col"
          >
            B
          </th>
          <th
            class="MuiTableCell-root Spreadsheet-tableCellRoot-5 MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
            scope="col"
          >
            C
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root Spreadsheet-tableBodyRoot-4 css-apqrd9-MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root Spreadsheet-tableRowRoot-6 css-1q1u3t4-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root Spreadsheet-tableCellRoot-5 MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
          >
            <div
              class="Spreadsheet-editBoxRoot-15 MuiBox-root css-0"
            >
              3
              <svg
                class="Spreadsheet-editIcon-16"
              >
                edit.svg
              </svg>
            </div>
          </td>
          <td
            class="MuiTableCell-root Spreadsheet-tableCellRoot-5 MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
          >
            <div
              class="Spreadsheet-editBoxRoot-15 MuiBox-root css-0"
            >
              18
              <svg
                class="Spreadsheet-editIcon-16"
              >
                edit.svg
              </svg>
            </div>
          </td>
          <td
            class="MuiTableCell-root Spreadsheet-tableCellRoot-5 MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
          >
            <div
              class="Spreadsheet-editBoxLast-17 Spreadsheet-editBoxRoot-15 MuiBox-root css-0"
            >
              13.5
              <svg
                class="Spreadsheet-editIcon-16"
              >
                edit.svg
              </svg>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
    <button
      class="MuiButtonBase-root MuiButton-root Spreadsheet-addRowButtonRoot-9 MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root Spreadsheet-addRowButtonRoot-9 MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
      tabindex="0"
      type="button"
    >
      Add row +
      <span
        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
